From d34dc7ba9c84fc323f279cb090823fd2853ac3b9 Mon Sep 17 00:00:00 2001
From: Hansong Zhang <hsz@google.com>
Date: Fri, 9 Feb 2018 14:16:59 -0800
Subject: [PATCH 4/6] DO NOT MERGE Truncate new line characters when adding
 string to config

Bug: 70808273
Test: test with a device with newline character in name
Change-Id: I8729e12ad5851ee1ffbcb7c08e9a659f768ffc21
(cherry picked from commit dd9bbfc2458569d9fecf35f7503d1b89b4c69aa0)
(cherry picked from commit 7f8bfcc35285ca6e93a4436699bc95c13b920caf)
mh0rst: Port to C
---
 osi/src/config.c | 26 ++++++++++++++++++++++++--
 1 file changed, 24 insertions(+), 2 deletions(-)

diff --git a/osi/src/config.c b/osi/src/config.c
index 345f907..b38977f 100644
--- a/osi/src/config.c
+++ b/osi/src/config.c
@@ -34,6 +34,7 @@
 #include "osi/include/allocator.h"
 #include "osi/include/list.h"
 #include "osi/include/log.h"
+#include "log/log.h"
 
 typedef struct {
   char *key;
@@ -216,16 +217,37 @@ void config_set_string(config_t *config, const char *section, const char *key, c
     list_append(config->sections, sec);
   }
 
+  char *value_string = (char *)value;
+  char *value_no_newline;
+  char *newline = strstr(value_string, "\n");
+  if (newline) {
+    android_errorWriteLog(0x534e4554, "70808273");
+    size_t newline_pos = newline - value_string;
+    value_no_newline = osi_strndup(value_string, newline_pos);
+    if (!value_no_newline) {
+      LOG_ERROR(LOG_TAG,"%s: Unable to allocate memory for value_no_newline", __func__);
+      return;
+    }
+  } else {
+    value_no_newline = value_string;
+  }
+
   for (const list_node_t *node = list_begin(sec->entries); node != list_end(sec->entries); node = list_next(node)) {
     entry_t *entry = list_node(node);
     if (!strcmp(entry->key, key)) {
       osi_free(entry->value);
-      entry->value = osi_strdup(value);
+      entry->value = osi_strdup(value_no_newline);
+      if (newline) {
+        osi_free(value_no_newline);
+      }
       return;
     }
   }
 
-  entry_t *entry = entry_new(key, value);
+  entry_t *entry = entry_new(key, value_no_newline);
+  if (newline) {
+    osi_free(value_no_newline);
+  }
   list_append(sec->entries, entry);
 }
 
-- 
2.7.4

